File: AirDropSend.swift
// AirDropSend.swift — open the AirDrop picker for the given files and stay alive
// until the transfer is done.
//
// Why a compiled app and not a script: NSSharingService's picker is a real
// AppKit window. It needs a process running a real NSApplication event loop to
// receive the click. `osascript` cannot provide one — and the failure is nasty
// rather than obvious: the picker DRAWS (the window server does that), but no
// mouse event is ever dispatched to it, so clicking a recipient produces no
// delegate callback at all, the window goes half-transparent instead of
// closing, and nothing is sent. Touching NSApplication from osascript to fix
// that suppresses the picker entirely. There is no script-shaped answer here.
import AppKit

let logURL = FileManager.default.homeDirectoryForCurrentUser
    .appendingPathComponent("Library/Caches/airdrop-send.log")

func note(_ m: String) {
    let stamp = ISO8601DateFormatter().string(from: Date())
    let line = "\(stamp)  \(m)\n"
    guard let data = line.data(using: .utf8) else { return }
    if let fh = try? FileHandle(forWritingTo: logURL) {
        defer { try? fh.close() }
        _ = try? fh.seekToEnd()
        try? fh.write(contentsOf: data)
    } else {
        try? data.write(to: logURL)
    }
}

final class ShareDelegate: NSObject, NSSharingServiceDelegate {
    func sharingService(_ s: NSSharingService, willShareItems items: [Any]) {
        note("willShareItems — picker presented")
    }
    // The transfer is asynchronous, so do not quit the instant this arrives.
    func sharingService(_ s: NSSharingService, didShareItems items: [Any]) {
        note("didShareItems — accepted by the service")
        DispatchQueue.main.asyncAfter(deadline: .now() + 3) { NSApp.terminate(nil) }
    }
    func sharingService(_ s: NSSharingService, didFailToShareItems items: [Any], error: Error) {
        let e = error as NSError
        // NSUserCancelledError (3072) is the user pressing Cancel: not a fault.
        if e.code == NSUserCancelledError {
            note("cancelled by user")
        } else {
            note("FAILED — \(e.domain) code=\(e.code) — \(e.localizedDescription)")
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) { NSApp.terminate(nil) }
    }
}

let paths = Array(CommandLine.arguments.dropFirst())
guard !paths.isEmpty else { note("no files given"); exit(0) }
let urls = paths.map { URL(fileURLWithPath: $0) }
note("--- invoked with \(urls.count) item(s) ---")
for u in urls { note("  \(u.path)") }

guard let svc = NSSharingService(named: .sendViaAirDrop) else {
    note("AirDrop sharing service unavailable"); exit(1)
}
note("canPerform = \(svc.canPerform(withItems: urls))")

let app = NSApplication.shared
let delegate = ShareDelegate()
svc.delegate = delegate
// Accessory: no Dock icon, no menu bar, but a real activatable app — which is
// exactly what the picker needs and what a script cannot be.
app.setActivationPolicy(.accessory)

// Hard ceiling, so a picker left open forever does not leave a process behind.
DispatchQueue.main.asyncAfter(deadline: .now() + 900) {
    note("timeout — quitting"); NSApp.terminate(nil)
}

DispatchQueue.main.async {
    app.activate(ignoringOtherApps: true)
    svc.perform(withItems: urls)
    note("perform(withItems:) called — picker should be on screen")
}

app.run()

File: airdrop-send.js
// airdrop-send.js — open the AirDrop picker preloaded with the given files.
//
// Run by the Finder Quick Action (see install-quick-action.sh):
//     osascript -l JavaScript airdrop-send.js <path> [<path> ...]
//
// Why JavaScript and not AppleScript: the sharing service reports failures
// through sharingService:didFailToShareItems:error:, and AppleScript CANNOT
// implement that selector — `error` is a reserved word and may not appear as a
// handler label. So an AppleScript version is structurally blind to the one
// callback that says why a send failed, which is exactly the information needed
// when the picker opens, you click a person, and nothing happens.
//
// ⚠️ Do NOT "improve" this by (a) calling setActivationPolicy: /
// activateIgnoringOtherApps:, or (b) compiling it into an .app. Both were tried
// on macOS 26.5.1 and both stop the picker from ever being drawn, while every
// log line still says success.
ObjC.import('Foundation');
ObjC.import('AppKit');

var LOG = ObjC.unwrap($.NSHomeDirectory()) + '/Library/Caches/airdrop-send.log';

function note(m) {
  try {
    var line = new Date().toISOString().replace('T', ' ').substr(0, 19) + '  ' + m + '\n';
    var path = $(LOG);
    var fm = $.NSFileManager.defaultManager;
    if (!fm.fileExistsAtPath(path)) {
      $(line).writeToFileAtomicallyEncodingError(path, true, $.NSUTF8StringEncoding, null);
    } else {
      // Keep it from growing without bound; this is a breadcrumb trail, not an archive.
      var attrs = fm.attributesOfItemAtPathError(path, null);
      if (attrs && ObjC.unwrap(attrs.objectForKey('NSFileSize')) > 200000) {
        fm.removeItemAtPathError(path, null);
        $(line).writeToFileAtomicallyEncodingError(path, true, $.NSUTF8StringEncoding, null);
        return;
      }
      var fh = $.NSFileHandle.fileHandleForWritingAtPath(path);
      fh.seekToEndOfFile;
      fh.writeData($(line).dataUsingEncoding($.NSUTF8StringEncoding));
      fh.closeFile;
    }
  } catch (e) { /* logging must never break sending */ }
}

ObjC.registerSubclass({
  name: 'AirDropSendDelegate',
  superclass: 'NSObject',
  methods: {
    'sharingService:willShareItems:': {
      types: ['void', ['id', 'id']],
      implementation: function (s, items) { note('willShareItems — recipient chosen, handoff starting'); }
    },
    'sharingService:didShareItems:': {
      types: ['void', ['id', 'id']],
      implementation: function (s, items) { note('didShareItems — service reports the send succeeded'); }
    },
    'sharingService:didFailToShareItems:error:': {
      types: ['void', ['id', 'id', 'id']],
      implementation: function (s, items, err) {
        try {
          note('didFailToShareItems — domain=' + ObjC.unwrap(err.domain) +
               ' code=' + err.code + ' — ' + ObjC.unwrap(err.localizedDescription));
        } catch (e) { note('didFailToShareItems (error unreadable): ' + e); }
      }
    }
  }
});

function run(argv) {
  if (argv.length === 0) { note('called with no files'); return 'no files'; }
  note('--- invoked with ' + argv.length + ' item(s) ---');

  var urls = $.NSMutableArray.array;
  for (var i = 0; i < argv.length; i++) {
    urls.addObject($.NSURL.fileURLWithPath($(argv[i])));
    note('  ' + argv[i]);
  }

  var svc = $.NSSharingService.sharingServiceNamed($.NSSharingServiceNameSendViaAirDrop);
  if (svc.isNil()) { note('AirDrop sharing service unavailable'); return 'no service'; }
  note('canPerformWithItems = ' + svc.canPerformWithItems(urls));

  // The delegate is held weakly by the service, so keep our own strong reference
  // or it is collected and the callbacks silently stop arriving.
  var del = $.AirDropSendDelegate.alloc.init;
  globalThis.__airdropDelegate = del;
  svc.delegate = del;

  svc.performWithItems(urls);
  note('performWithItems returned — picker should be on screen');

  // Outlive the transfer. The picker belongs to THIS process and dies with it,
  // and there is deliberately no early exit on didShareItems: that callback
  // fires when the recipient is PICKED, not when the file has landed.
  var rl = $.NSRunLoop.currentRunLoop;
  for (var k = 0; k < 3600; k++) {
    rl.runUntilDate($.NSDate.dateWithTimeIntervalSinceNow(0.25));
  }
  note('--- hold window ended ---');
  return 'done';
}
